feat: Add RAGAS evaluation framework for RAG quality assessment

anouar-bm opened the pull request:
## Description

This PR adds a comprehensive evaluation system using the RAGAS framework to assess LightRAG's retrieval and generation quality. The evaluation framework provides automated testing and quality metrics for RAG systems, enabling developers to measure and improve their RAG implementations.

## Related Issues

Addresses the need for systematic RAG quality assessment and provides tools for continuous monitoring of LightRAG's performance.

## Changes Made

### New Files Added
- **`lightrag/evaluation/eval_rag_quality.py`** (394 lines)
  - `RAGEvaluator` class with async evaluation support
  - Four key RAG quality metrics:
    - **Faithfulness** (0.0-1.0): Measures answer factual accuracy vs retrieved context
    - **Answer Relevance** (0.0-1.0): Evaluates query-response alignment
    - **Context Recall** (0.0-1.0): Assesses retrieval completeness
    - **Context Precision** (0.0-1.0): Measures retrieved context quality/noise ratio
  - HTTP API integration for testing live LightRAG instances
  - JSON and CSV report generation with timestamps
  - Configurable test datasets via JSON files
  - Async evaluation with progress tracking

- **`lightrag/evaluation/README.md`** (309 lines)
  - Complete installation guide
  - Usage examples with code samples
  - Detailed metric interpretation
  - Troubleshooting section
  - Best practices for RAG evaluation

- **`lightrag/evaluation/__init__.py`** (16 lines)
  - Package initialization with version info

### Modified Files
- **`pyproject.toml`**
  - Added optional `evaluation` dependencies group:
    - `ragas>=0.3.7` (RAG assessment framework)
    - `datasets>=4.3.0` (HuggingFace datasets for RAGAS)
    - `httpx>=0.28.1` (HTTP client for API testing)
    - `pytest>=8.4.2` (testing framework)
    - `pytest-asyncio>=1.2.0` (async test support)

- **`.gitignore`**
  - Added `lightrag/evaluation/results/` to exclude generated evaluation reports

## Installation

```bash
# Install with evaluation dependencies
pip install lightrag-hku[evaluation]
```

## Usage Example

```python
from lightrag.evaluation.eval_rag_quality import RAGEvaluator

# Initialize evaluator
evaluator = RAGEvaluator(
    api_base_url="http://localhost:8000",
    test_dataset_path="lightrag/evaluation/test_dataset.json",
    output_dir="lightrag/evaluation/results"
)

# Run evaluation
results = await evaluator.evaluate_all()

# View results
print(f"Average Faithfulness: {results['avg_faithfulness']:.2f}")
print(f"Average Answer Relevance: {results['avg_answer_relevance']:.2f}")
print(f"Average Context Recall: {results['avg_context_recall']:.2f}")
print(f"Average Context Precision: {results['avg_context_precision']:.2f}")
```

## Checklist

- [x] Changes tested locally with sample datasets
- [x] Code reviewed for quality and best practices
- [x] Documentation updated (comprehensive README included)
- [x] Unit tests added (evaluation framework with test dataset)
- [x] Dependencies properly declared as optional

## Additional Notes

### Design Principles
This evaluation framework is designed to be:
- **Generic**: Works with any LightRAG instance via HTTP API
- **Non-intrusive**: Zero changes to core LightRAG code
- **Optional**: Installed separately via `pip install lightrag-hku[evaluation]`
- **Extensible**: Easy to add custom metrics or test datasets
- **Production-ready**: Async evaluation, error handling, progress tracking

### Why RAGAS?
RAGAS is a widely-adopted framework in the RAG community that provides:
- Standardized metrics for comparing RAG systems
- Research-backed evaluation methodology
- Active maintenance and community support
- Integration with popular ML frameworks

### Use Cases
- **Development**: Measure improvements during RAG system development
- **CI/CD**: Automated quality checks in deployment pipelines
- **A/B Testing**: Compare different RAG configurations
- **Monitoring**: Track RAG quality over time in production

### Performance
- Async evaluation for concurrent test execution
- Caching support to avoid re-processing
- Configurable batch sizes
- Progress tracking for long-running evaluations

Thank you for reviewing this contribution!

danielaskdd commented:
Regarding the query interface's context return, please consider the following improvements:  
1. Since a single file may correspond to multiple chunks, the `content` field should be a list. Please update the documentation and examples accordingly to reflect this.  
2. The `query_text_stream` parameter should support the `include_references` option, just like `query_text`, to enable the client to receive reference information in the response.

danielaskdd commented:
@codex review

anouar-bm commented:
## Updates

✅ Changed `content` field from string to `List[str]` with `ReferenceItem` Pydantic model
✅ Added `include_chunk_content` support to `/query/stream` endpoint
✅ Fixed evaluation module: moved `sys.exit()` to runtime, added lazy imports, updated docs

**Note:** Test data in `sample_dataset.json` is generic (not personal) - users should customize based on their indexed documents.

anouar-bm commented:
@danielaskdd The evaluation results I shared (97.37% RAGAS score) used my personal portfolio documents with **customized entity extraction prompts** for better results. Neither the documents nor custom prompts are in this repo.

The current `sample_dataset.json` contains generic LightRAG questions as a template. I can add matching sample markdown files (covering LightRAG features, architecture, vector DBs, deployment) so users can reproduce evaluation results with the default prompts. Should I create these?

anouar-bm commented:
### ✅ Added Sample Documents for Reproducible Testing

I've added 5 markdown documents that users can index to reproduce evaluation results.

**Changes:**
- Added `sample_documents/` folder with 5 markdown files covering LightRAG features
- Updated `sample_dataset.json` with 3 improved, specific test questions
- Shortened and corrected evaluation README (removed outdated info about mock responses)
- Added sample_documents reference with expected ~95% RAGAS score

**Test Results with sample documents:**
- Average RAGAS Score: **95.28%**
- Faithfulness: **100%**, Answer Relevance: **96.67%**
- Context Recall: **88.89%**, Context Precision: **95.56%**

Users can now index these docs and reproduce the benchmark results.

danielaskdd commented:
# Enhanced RAGAS Evaluation Framework

The updates significantly enhance **compatibility** with any OpenAI-compatible service (including custom endpoints via `bypass_n` mode) and introduce **intelligent concurrency control** to prevent rate limiting. Additionally, the update features **robust NaN handling**, **comprehensive configuration options**, and **professional formatted output**.

### Before:

- **Compatibility issue with RAGAS 0.3.x, specifically concerning the `answer_relevancy` metric's default strictness parameter.**
- **No concurrency control (frequent rate limiting)**
- **Race condition may result in the injected LLM being set to None when two evaluations overlap**
- Basic RAGAS integration with hard-coded OpenAI dependency
- Insufficient error handling for NaN values during average calculation
- Minimal configuration options
- Does not support LightRAG Server when API key authentication is required.

### After:

- **Full compatibility with custom OpenAI-compatible endpoints** via `bypass_n` mode
- **Intelligent concurrency control to prevent rate limiting**
- **RAGAS run now builds fresh metric objects** (Faithfulness(), AnswerRelevancy(), ContextRecall(), ContextPrecision()).
- Full support for any OpenAI-compatible service ( LM Studio, vLLM, etc.)
- Flexible model configuration with fallbacks
- Robust NaN handling across all metrics calculation
- Comprehensive environment-based configuration
- Support LIGHTRAG_API_KEY auth header
- Professional formatted output
- **Enviroment variable added** (env.example)

| Variable                   | Default                          | Purpose                   |
| -------------------------- | -------------------------------- | ------------------------- |
| `EVAL_LLM_MODEL`           | `gpt-4o-mini`                    | LLM for RAGAS evaluation  |
| `EVAL_EMBEDDING_MODEL`     | `text-embedding-3-small`         | Embedding model           |
| `EVAL_LLM_BINDING_API_KEY` | (falls back to `OPENAI_API_KEY`) | API key                   |
| `EVAL_LLM_BINDING_HOST`    | (optional)                       | Custom endpoint URL       |
| `EVAL_MAX_CONCURRENT`      | `1`                              | Concurrent evaluations    |
| `EVAL_QUERY_TOP_K`         | `10`                             | Documents per query       |
| `EVAL_LLM_MAX_RETRIES`     | `5`                              | Request retries           |
| `EVAL_LLM_TIMEOUT`         | `120`                            | Request timeout (seconds) |

* Renamed `"context"` field to `"project"`  in Sample Dataset (sample_dataset.json)

## 🔧 Technical Details

1. **RAGAS Compatibility**: RAGAS 0.3.x requires properly wrapped LLM instances
2. **Custom Endpoint Support**: Many OpenAI-compatible services don't support all OpenAI API parameters
3. **The `n` Parameter Problem**:
   - OpenAI's API supports `n` parameter to generate multiple completions in one request
   - Custom endpoints (Ollama, LM Studio, vLLM, etc.) often reject this parameter
   - RAGAS metrics like `answer_relevancy` internally use `n` for generating multiple variants
4. **The Solution**: `bypass_n=True` makes RAGAS generate multiple outputs through repeated single prompts instead

danielaskdd commented:
Thank you for your contribution to LightRAG. It has been a pleasure collaborating with you.

anouar-bm commented:
I will push later the following workflow that I use:

```mermaid
flowchart TD
    A[Query + Conversation History] --> B[Backend: Check if query needs RAG]
    
    B -->|Needs RAG| C[Return: need_rag = yes; query_enhanced = string]
    C --> D[Query becomes independent of conversation history]
    D --> E[Process RAG: Retrieve relevant documents]
    E --> F[Generate response with RAG context]
    F --> G[Return RAG response]
    
    B -->|Does NOT need RAG| H[Return: need_rag = no; query_response = string]
    H --> I[Return query_response directly]
    
    G --> J[Final Response to User]
    I --> J[Final Response to User]
